src/board.js

```diff
@@ -9,22 +9,232 @@
+        points.join(", ") + " with liberties " + liberties.join(", ") + ")";
+    },
+  };
+
   // options:
   //   - size: typically 19.
   //   - komi: floating-point number.
   function Board(options) {
     options = options | {};
     this.size = +options.size || 19;
     this.komi = +options.komi || 7.5;
-    this.board = new Uint8Array(this.size * this.size);
+    var numPoints = this.size * this.size;
+    this.board = new Array(numPoints);
+    for (var y = 0; y < this.size; y++) {
+      for (var x = 0; x < this.size; x++) {
+        this.board[x + y * this.size] = new Point(x, y);
+      }
+    }
+    this.groups = new Set();
+    this.nextPlayingColor = Board.BLACK;
   }
 
   Board.prototype = {
     // x, y: integer positions of an intersection on the board.
-    get(x, y) { return this.board[x + y * this.size]; },
-    set(x, y, color) { return this.board[x + y * this.size] = color; },
+    get: function(x, y) { return this.board[x + y * this.size]; },
+    set: function(x, y, color) { this.board[x + y * this.size].color = color; },
+
+    mergeGroups: function(groups) {
+      var points = [];
+      groups.forEach(function(group) {
+        group.points.forEach(function(point) {
+          points.push(point);
+        });
+      });
+      return new Group(this, points);
+    },
+
+    surrounding: function(x, y) {
+      var top    = (y > 0)?             this.get(x    , y - 1): undefined;
+      var right  = (x < this.size - 1)? this.get(x + 1, y    ): undefined;
+      var bottom = (y < this.size - 1)? this.get(x    , y + 1): undefined;
+      var left   = (x > 0)?             this.get(x - 1, y    ): undefined;
+      return [top, right, bottom, left];
+    },
+
+    // Place a stone.
+    play: function(x, y) {
+      var self = this;
+      var point = self.get(x, y);
+      var color = self.nextPlayingColor;
+      if (point === undefined || point.color !== Board.EMPTY) { return false; }
+      self.set(x, y, color);
+
+      // Get surrounding groups.
+      var surrounding = self.surrounding(x, y);
+      var surroundingGroups = surrounding.map(function(neighbor) {
+        if (neighbor === undefined) { return; }  // Off the board.
+        if (neighbor.color === Board.EMPTY) { return; }
+        return neighbor.group;
+      }).filter(function(group) { return group !== undefined; });
+      var ownSurroundingGroups = surroundingGroups.filter(function(group) {
+        return group.color === color;
+      });
+      var enemySurroundingGroups = surroundingGroups.filter(function(group) {
+        return group.color !== color;
+      });
+      var capturedEnemyGroups = enemySurroundingGroups.filter(function(group) {
+        return group.liberties.size === 1;
+      });
+
+      if (capturedEnemyGroups.length === 0) {
+        // We are not capturing enemy stones. Are we committing suicide?
+        var isDirectlySurroundedByEnemy = enemySurroundingGroups.length === 4;
```

Comment:
This doesn't look right, I'll fix it. There can be a single enemy group surrounding the point where we try to place a stone.